File: 프로그래머스/3/43163. 단어 변환/단어 변환.java
class Solution {
    public static int answer = Integer.MAX_VALUE;
    
    public int solution(String begin, String target, String[] words) {
        boolean[] visited = new boolean[words.length];
        
        dfs(begin, target, words, visited, 0);
        
        if(answer == Integer.MAX_VALUE) {
            answer = 0;
        }
        
        return answer;
    }
    
    public void dfs(String begin, String target, String[] words, boolean[] visited, int count) {
        if(begin.equals(target)) {
            answer = Math.min(answer, count);
            return;
        }
        
        for(int i = 0; i < words.length; i++) {
            if(visited[i]) {
                continue;
            }
            
            if(canChange(begin, words[i])) {
                visited[i] = true;
                dfs(words[i], target, words, visited, count+1);
                visited[i] = false;
            }
        }
    }
    
    public boolean canChange(String begin, String target) {
        int count = 0;
        if(begin.equals(target)) {
            return false;
        }
        
        for(int i = 0; i < begin.length(); i++) {
            if(begin.charAt(i) != target.charAt(i)) {
                count++;
            }
        }
        
        if(count == 1) {
            return true;
        }
        
        return false;
    }
}

File: 프로그래머스/3/43163. 단어 변환/단어 변환.js
function solution(begin, target, words) {
  let answer = Number.MAX_SAFE_INTEGER;
  const visited = new Array(words.length).fill(false);

  function dfs(current, count) {
    if (current === target) {
      answer = Math.min(answer, count);
      return;
    }

    for (let i = 0; i < words.length; i++) {
      if (!visited[i] && canChange(current, words[i])) {
        visited[i] = true;
        dfs(words[i], count + 1);
        visited[i] = false;
      }
    }
  }

  function canChange(word1, word2) {
    let diff = 0;
    for (let i = 0; i < word1.length; i++) {
      if (word1[i] !== word2[i]) diff++;
    }
    return diff === 1;
  }

  dfs(begin, 0);

  return answer === Number.MAX_SAFE_INTEGER ? 0 : answer;
}